omniscidb/Calcite/java/calcite/src/main/java/org/apache/calcite/rel/externalize/MapDRelJson.java

```diff
@@ -518,6 +533,73 @@ RexNode toRex(RelInput relInput, Object o) {
     }
   }
 
+  static TimestampString longToTimestampString(long value, int precision) {
+    assert (precision == 0) || (precision == 3) || (precision == 6) || (precision == 9);
+
+    long pow = (long) Math.pow(10, precision);
+    long seconds = value / pow;
+    LocalDateTime dt = LocalDateTime.ofEpochSecond(seconds, 0, ZoneOffset.UTC);
+    int year = dt.getYear();
+    assert year < 10000;
+    int month = dt.getMonthValue();
+    int day = dt.getDayOfMonth();
+    int hour = dt.getHour();
+    int minute = dt.getMinute();
+    int second = dt.getSecond();
+
+    long fraction = value - (seconds * pow);
+    for (; (precision > 0) && ((fraction % 10) == 0); precision--, fraction /= 10) ; // Cut trailing zeros
+
+    char[] ts = new char[precision == 0 ? 19 : 20 + precision];
+    ts[0] = (char) ('0' + (year / 1000));
+    ts[1] = (char) ('0' + ((year / 100) % 10));
+    ts[2] = (char) ('0' + ((year / 10) % 10));
+    ts[3] = (char) ('0' + (year % 10));
+    ts[4] = '-';
+    ts[5] = (char) ('0' + (month / 10));
+    ts[6] = (char) ('0' + (month % 10));
+    ts[7] = '-';
+    ts[8] = (char) ('0' + (day / 10));
+    ts[9] = (char) ('0' + (day % 10));
+    ts[10] = ' ';
+    ts[11] = (char) ('0' + (hour / 10));
+    ts[12] = (char) ('0' + (hour % 10));
+    ts[13] = ':';
+    ts[14] = (char) ('0' + (minute / 10));
+    ts[15] = (char) ('0' + (minute % 10));
+    ts[16] = ':';
+    ts[17] = (char) ('0' + (second / 10));
+    ts[18] = (char) ('0' + (second % 10));
```

Comment:
I'm a bit confused about all this part. Are there no methods in java to cut trailing zeroes or string builders?

Reply:
There is the `StringBuilder` class, but I believe it would not be more efficient for this use case, because here know exactly the length of the string and positions of each character.